docs/RFCS/20190906_flow_control.md

```diff
@@ -0,0 +1,137 @@
+- If concurrency doesn't change, but TPS decreases and latency increases significantly (maybe TPS halves and latency doubles), we can believe the congestion is occurring.
+- If concurrency doesn't change, TPS and latency don't change too much, but data loss rate increases significantly (however in most cases if data loss rate increases, we often find TPS decreases and latency increases), we can believe the congestion is occurring.
+
+The congestion degree can be measured by the three indicators, but the workflow in the real world is often complicated, we won't introduce an estimate model in this proposal, as we only need to known whether the congestion is occurring in our congestion control model.
+
+### Congestion control framework
+
+The congestion control framework is separated into three parts, the congestion detect component, the data flow control component and the concurrency control component.
+
+#### The congestion detect component
+
+The congestion detect component will collect the key indicators in DM-worker as following:
+
+- data loss rate: we will add a sliding window to record the count of downstream resource busy error (such as `tmysql.ErrPDServerTimeout`, `tmysql.ErrTiKVServerTimeout`, `tmysql.ErrTiKVServerBusy` etc) and the count of all dispatched transactions. The data loss rate will be calculated with these two counters.
+- TPS detection: we will add a sliding window to record the count of successfully executed transactions, and calculate TPS with this value.
+- latency: transaction execution latency is already recorded in load and sync unit, we can use it directly.
+
+#### The data flow control component
+
+This component is responsible to give a reasonable concurrency value based on the data flow stage, current and historical data flow indicators. Firstly we define some stage, which is used to represent the concurrency change stage.
+
+```go
+// CongestionStage represents the stage in a congestion control algorithm
+type CongestionStage int
+
+// Congestion stage list
+const (
+    // When the task starts with a low concurrency, the concurrency can be increased fast in this stage.
+    StageSlowStart CongestionStage = iota + 1
+    // When the concurrency exceeds a threshold, we recommend to increase slowly, which is often called the avoidance stage.
+    StageAvoidance
+    // The steady stage means the data flow TPS reaches the bandwidth, the latency keeps steady and no data loss.
+    StageSteady
+    // When congestion happens and the concurrency is turned down, we enter the recovery stage.
+    StageRecovery
+)
+```
+
+We define a `FlowControl` interface, which is used to give a reasonable concurrency value based on the congestion stage, the historical and current data flow indicators. We can use different congestion algorithm, just to implement the `FlowControl` interface.
```

Comment:
`which is used to give a reasonable concurrency value based on the congestion stage, the historical and current data flow indicators` This description is a bit repetitive, see L96